[#91] MySql Replication 적용에 따른 DataSource 변환 및 쿼리 분산

hanwix2 opened the pull request:
## 개요
- DB의 **부하 분산**과 **백업**을 위한 **DB Replication** 적용
    - 같은 로컬 서버에 서로 다른 port를 사용하여 Master(3306)와 Slave(3307) DB 서버 구축
- DB 쿼리 종류에 따라 DataSource 변환 적용
    - **Master**: DB에 쓰기, 수정, 삭제 쿼리문 수행
    - **Slave**: DB 읽기 쿼리문 수행


## Progress
- ~~Annotation(`DatabaseSource`)을 이용한 AOP를 적용~~
    - ~~`DatabaseSourceAspect` - mapper의 DB 처리 메소드 수행 전 DB Source 결정~~
- ~~`ThreadLocal`에 같은 작업(스레드)내에서 value(`DatabaseSource`)를 저장 & 사용(DB source 결정용)~~
- `AbstractRoutingDataSource`를 이용해 annotation의 value(`DatabaseSource`)에 따라 DataSource 연결
- **Read-Only** 트랜잭션의 작업은 **Slave server**에서 처리
        - `TransactionSynchronizationManager`를 통해 해당 트랜잭션이 read-only 인지 확인
- `LazyConnectionSourceProxy`:
        - 스프링에서 트랜잭션이 시작할 때 connection을 먼저 확보하는 것을 방지하기 위해 사용
        - 만약 Connection이 먼저 확보된다면 기본값인 Master만 사용 됨

### etc
~~다중 Datasource를 적용하기 위해 **Transaction** 을 이용한 방식이 있지만,~~
~~불필요하게 과도한 @Transaction 어노테이션 사용과 메소드간 DB 작업이 중첩될 시 복잡함이 야기될 수 있기에 **AOP** 방식을 적용.~~
> ~~**Transaction** 방식은 트랜잭션의 ReadOnly 여부에 따라 master/slave 분기하는 방법이 있다.~~
- 기존 AOP 기반 방식은 DataSource 관련 관리 포인트가 더 추가되므로 관리비용의 증가라는 단점 존재
- 대부분의 서비스 로직에는 트랜잭션이 적용되어야 하므로 관리의 효율을 위해 트랜잭션 기반으로 DataSource 분기 방식으로 변경

hanwix2 committed:
[#91] MySql Replication 적용에 따른 DB 서버 분기

- 개요:
    - DB의 부하 분산과 백업을 위한 DB Replication 적용(로컬의 DB에 Master와 Slave 서버 구축)
    - 서버 분기에 따라 상황에 맞게 DataSource 변환 적용
        - Master: DB에 쓰기, 수정, 삭제 쿼리문 수행
	- Slave: DB 읽기 쿼리문 수행
- 구현:
    - Annotation(DatabaseSource)을 이용한 AOP를 적용하여 mapper의 DB 처리 메소드 수행 전 DB Source 결정 - DatabaseSourceAspect
    - AbstractRoutingDataSource를 이용해 annotation의 value(DatabaseSource)에 따라 DataSource 연결
    - ThreadLocal에 같은 작업(스레드)내에서 value(DatabaseSource)를 저장 & 사용(DB source 결정용)

tlsgkr7416 commented:
개요에 보면 master 서버에서는 읽기 기능을 처리하지 않는 거 같은데 만약 user에게 변화된 데이터를 빠르게 조회해야 될 경우 master 서버에서 읽기 기능을 처리해야 되지 않을까요? 왜냐면 slave 서버는 비동기로 실시간 데이터가 복제되어 데이터 동기화가 늦어질 수 있을 거 같아서요.  **그래서 제 생각은 읽기 기능을 천천히 조회해도 되는 데이터는 slave 서버에 빠르게 조회해야 되는 데이터는 master서버에 설정해야 된다고 생각해요.**  저도 확실한 게 아니여서 피드백 받아보고 싶습니다.

tlsgkr7416 commented on src/main/java/com/show/showticketingservice/mapper/PerformanceMapper.java:
```diff
@@ -1,30 +1,41 @@
 @Mapper
 public interface PerformanceMapper {
 
+    @DatabaseSource(DatabaseServer.MASTER)
     void insertPerformance(PerformanceRequest performanceRequest);
 
+    @DatabaseSource(DatabaseServer.SLAVE)
     boolean isPerformanceExists(String title, ShowType showType);
 
```
여기는 트랜잭션으로 묶여 있는데 서버 한곳에서 처리해야 되는 거 아닌가요?

f-lab-dev commented on src/main/java/com/show/showticketingservice/mapper/PerformanceMapper.java:
```diff
@@ -1,30 +1,41 @@
+    @DatabaseSource(DatabaseServer.SLAVE)
     boolean isPerformanceIdExists(int performanceId);
 
+    @DatabaseSource(DatabaseServer.SLAVE)
```
이렇게 일일이 붙여주는건 너무 관리비용이 커지지 않을까요?
read-only 트랜잭션에만 슬레이브에서 읽도록 하면 좋을 것 같아요~
이런 귀찮은 노가다성 작업이 늘어난다면 이걸 줄일 수 있는 방법을 고민해보시는 것도 방법입니다~

hanwix2 replied:
기본 DataSource는 Master이기 때문에 **Slave 서버에 요청하는 쿼리만 어노테이션을 달고** Master는 달지 않아도 될 것 같은데 
일단 제가 기능 구현 시 명확히 나누기 위해 전부 붙여두긴 했습니다.

f-lab-dev reviewed (commented):
신학님이 말씀하신대로 실시간 읽기가 필요한 곳은 마스터에서 읽으면 좋습니다~

hanwix2 commented:
Spring에서는 트랜잭션이 시작되기 전 DataSource를 미리 확보를 해서 동일한 DataSource에서 트랜잭션이 진행된다고 합니다. 
이는 제가 구현한 (DB 쿼리문이 실행되기 전) AOP로 DataSource를 지정하기 전 
**트랜잭션이 시작될 때 이미 기본 DataSource인 Master server로 연결**이 됩니다.

- 신학님이 말씀하신 "빠르게 조회해야 되는 데이터" 는 트랜잭션 내에 있는 읽기 연산이라고 생각되는데 앞선 설명에 따르면 문제가 없다고 판단했습니다.  (만약 **티켓 예매 여부에 대한 읽기 연산**은 추후에 기능 구현 시 해당 쿼리를 Master에서 처리하도록 지정하게 하면 될 듯 하네요. 이외의 읽기 작업 중 빠르게 조회해야하는 데이터는 딱히 생각나지 않습니다.)

- 멘토님께서 Read-only 트랜잭션에 대해 Slave 서버를 통해 작업을 하라고 하셨는데,
저희 프로젝트에서는 아직 읽기 연산만 수행하는 트랜잭션이 없어 구현을 하진 않았습니다. 
(그리고 사실 Read-only 트랜잭션이 필요하다고 느낀 기능이 아직 없기도 한 것 같고요.) 
하지만 관리의 효율성을 위해서 미리 만들어 두는 것도 좋은 방법일 것 같습니다.

- Mapper의 DB 쿼리 관련 메소드에 하나하나 annotation을 붙여준 것은 이 메소드가 다른 서비스를 구현할 때 단일 작업으로 재사용 할 경우 미리 DataSource를 지정함으로 오히려 관리상의 편리함이 있진 않을까 생각합니다. 


혹시 제가 잘못 판단한 게 있다면 피드백 부탁드려요.
_(Read-only 트랜잭션에 대한 slave server 매칭 작업은 빠른시일 내에 적용하겠습니다!)_

---

**멘토링 시간 중 받은 피드백 적용**
- 트랜잭션 기반 DataSource 분기

hanwix2 committed:
[#91] DataSource 분기 방식을 트랜잭션 기반으로 변경

- 개요:
    - 기존 AOP 기반 방식은 DataSource 관련 관리 포인트가 더 추가되므로 관리비용의 증가라는 단점 존재
    - 대부분의 서비스 로직에는 트랜잭션이 적용되어야 하므로 관리의 효율을 위해
      트랜잭션 기반으로 DataSource 분기 방식으로 변경
- 구현:
    - Read-Only 트랜잭션의 작업은 Slave server에서 처리
        - TransactionSynchronizationManager를 통해 해당 트랜잭션이 read-only 인지 확인
    - LazyConnectionSourceProxy:
        - 스프링에서 트랜잭션이 시작할 때 connection을 먼저 확보하는 것을 방지하기 위해 사용
        - 만약 Connection이 먼저 확보된다면 기본값인 Master만 사용 됨
- etc:
    - 서비스 로직에 @Transactional 어노테이션 명시

f-lab-dev reviewed (commented):
👍

hanwix2 committed:
Merge branch 'develop' into feature/91/jh
